src/riscv/lib/src/jit/builder/instr_map.rs

```diff
@@ -0,0 +1,233 @@
+            .map(|(idx, value)| (InstrId(idx), value))
+    }
+
+    /// Transform the values in the `InstrMap` using the provided function, producing a new `InstrMap`.
+    pub(crate) fn map<F, U>(self, f: F) -> InstrMap<U>
+    where
+        F: Fn(InstrId, &V) -> U,
+    {
+        let instructions = self
+            .iter()
+            .map(|(instr_id, val)| f(instr_id, val))
+            .collect();
+        InstrMap { instructions }
+    }
+}
+
+impl<V> Index<InstrId> for InstrMap<V> {
+    type Output = V;
+
+    fn index(&self, index: InstrId) -> &Self::Output {
+        &self.instructions[index.0]
+    }
+}
+
+impl<V> IndexMut<InstrId> for InstrMap<V> {
+    fn index_mut(&mut self, index: InstrId) -> &mut Self::Output {
+        &mut self.instructions[index.0]
+    }
+}
+
+/// Builder for initialising an ([`AddrMap`], [`InstrMap`]) pair.
+pub(crate) struct InstrMapBuilder<T> {
+    addresses: Vec<Address>,
+    instructions: Vec<T>,
+}
+
+impl<T> InstrMapBuilder<T> {
+    /// Create a new builder with the specified capacity.
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            addresses: Vec::with_capacity(capacity),
+            instructions: Vec::with_capacity(capacity),
+        }
+    }
+
+    /// Insert a new address and instruction pair to the builder. If the address
+    /// already exists, override the existing instruction with the new instruction.
+    pub(crate) fn insert(&mut self, address: Address, instruction: T) {
+        // insert the address and instruction at the same index by binary searching the address list
+        match self.addresses.binary_search(&address) {
+            Ok(idx) => {
+                self.instructions[idx] = instruction;
+            }
+            Err(idx) => {
+                self.addresses.insert(idx, address);
+                self.instructions.insert(idx, instruction);
+            }
+        }
+    }
+
+    /// Finalise the builder and produce the (`AddrMap`, `InstrMap`) pair.
+    pub(crate) fn build(self) -> (AddrMap, InstrMap<T>) {
+        let instr_map = InstrMap {
+            instructions: self.instructions,
+        };
+        let addr_map = AddrMap {
+            addresses: self.addresses,
+        };
+
+        (addr_map, instr_map)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use proptest::proptest;
+
+    use super::*;
+
+    proptest! {
+        #[test]
+        fn addr_map_translate_and_address_of_are_inverses(addresses in proptest::collection::vec(0u64..1_000, 1..50)) {
+            let mut sorted_addresses = addresses.clone();
+            sorted_addresses.sort_unstable();
+            sorted_addresses.dedup();
+
+            let addr_map = AddrMap { addresses: sorted_addresses.clone() };
+
+            for (idx, &address) in sorted_addresses.iter().enumerate() {
+                let instr_id = addr_map.translate(address).expect("Address should translate to an InstrId");
+                assert_eq!(instr_id.0, idx);
```

Comment:
[`prop_assert`](https://docs.rs/proptest/latest/proptest/macro.prop_assert.html) is better in property-based tests. It makes the test output more understandable by skipping Rust's default panic handler, for example when shrinking the inputs to find a minimal counter example. This involves running the proptest multiple times which cause a panic each time and therefore print the panic.